pip/pip-376.md

```diff
@@ -0,0 +1,138 @@
+
+As you can see, these two methods are hard to use. And they are also only used in tests except for the `getTopicPoliciesAsyncWithRetry` method, which uses a user-provided executor and backoff policy to call `getTopicPolicies` until `TopicPoliciesCacheNotInitException` is not thrown:
+
+```java
+    default CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsyncWithRetry(TopicName topicName,
+              final Backoff backoff, ScheduledExecutorService scheduledExecutorService, boolean isGlobal) {
+```
+
+The `getTopicPolicies` overrides are only called in tests while `getTopicPoliciesAsyncWithRetry` is used in the core. It would be very confusing to users that want to implement their own topic policies service. They have to look deeply into the Pulsar's source code to know these details.
+
+https://github.com/apache/pulsar/pull/21231 adds two asynchronous overrides that are much more friendly to users:
+
+```java
+CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsync(@Nonnull TopicName topicName, boolean isGlobal);
+CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsync(@Nonnull TopicName topicName);
+```
+
+Now we have **5** asynchronous get methods. What's worse, unlike `getTopicPolicies`, `getTopicPoliciesAsync(topic)` is not equivalent to `getTopicPoliciesAsync(topic, false)`, instead,
+- `getTopicPoliciesAsync(topic)` will try getting local policies first, if absent, then try getting global policies
+- `getTopicPoliciesAsync(topic, true)` will try getting global policies
+- `getTopicPoliciesAsync(topic, false)` will try getting local policies
+
+It should be noted that the topic policies support global policies across clusters since [#12517](https://github.com/apache/pulsar/pull/12517). So there are local policies and global policies.
+
+Currently,
+- `getTopicPoliciesAsync(TopicName)` is used in `BrokerService#getTopicPoliciesBypassSystemTopic`, which is called when initializing the topic policies of `PersistentTopic` objects. So it uses the "local-first" semantics in case the global policies or local policies is deleted.
+- `getTopicPoliciesAsyncWithRetry` is used in `AdminResource#getTopicPoliciesAsyncWithRetry`, which is called for all topic policies admin APIs. Since these admin APIs all have a `isGlobal` field to indicate whether to get the global policies, it uses the "local only" or "global only" semantics.
+- Other methods are never called directly other than tests.
+
+Actually there is the 6th method `getTopicPoliciesIfExists`, which just tries to get the local topic policies from the cache. This method is the most clear and simple in all these stuffs.
+
+```java
+    TopicPolicies getTopicPoliciesIfExists(TopicName topicName);
+```
```

Comment:
This one is also confusing for plugin. If there is no policies for a topic, user can also implement by 

```
CompletableFuture<Optional<TopicPolicies>> getTopicPoliciesAsync(@Nonnull TopicName topicName, boolean isGlobal);
```

I remember it's also added for resolving the issues from SystemTopic based solution which don't want to wait for the message replay from the system topic.

Reply:
Let me recheck the references of this method.